frontend/src/textToSpeech/textToSpeech.js

```diff
@@ -0,0 +1,383 @@
+            showModal: false,
+            userText: '',
+            graded: false,
+            grade: null,
+            continue: false,
+            showAnswer: false,
+        };
+        this.modalHandler = this.modalHandler.bind(this);
+    }
+
+    componentDidMount = async () => {
+        const apiURL = `/api/get_indiv_sentences/${this.props.textID}`;
+        const response = await fetch(apiURL);
+        const data = await response.json();
+        this.setState({
+            textData: data,
+        });
+        console.log(this.state.textData);
+        document.addEventListener('keydown', this.handleKeyDown, true);
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener('keydown', this.handleKeyDown, false);
+    }
+
+    getCurrentSentence = () => {
+        return this.state.textData[this.state.sentenceIndex]['sentence'];
+    }
+
+    changeSentence = (delta) => {
+        const { textData, sentenceIndex } = this.state;
+        const listLength = textData.length;
+        const newsentenceIndex = (sentenceIndex + delta + listLength) % listLength;
+        this.setState({
+            sentenceIndex: newsentenceIndex,
+            showNext: true,
+        });
+        this.reset();
+    }
+
+    reset = async () => {
+        this.setState({
+            showModal: false,
+            userText: '',
+            graded: false,
+            grade: null,
+            continue: false,
+            showAnswer: false,
+        });
+        document.getElementById('content').value = '';
+    }
+
+    giveUp = () => {
+        this.setState({
+            continue: true,
+            showAnswer: true,
+        });
+    }
+
+    playAudio = () => {
+        const utterance = new SpeechSynthesisUtterance();
+        utterance.text = this.getCurrentSentence();
+        utterance.lang = 'en-US';
+        utterance.rate = 1;
+        speechSynthesis.speak(utterance);
+    }
+
+    handleSubmit = () => {
+        const input = document.getElementById('content').value;
+        console.log('this is the input i want', input);
+        this.setState({
+            userText: input,
+        });
+        this.gradeText(input);
+
+        console.log('You have submitted!!!');
+    };
+
+    gradeText = async (input) => {
+        console.log('gradeText was called');
+        try {
+            console.log('im trying something here');
+            const apiURL = `/api/get_sentence_grade/${input}/${this.getCurrentSentence()}`;
+            const response = await fetch(apiURL);
+            const grade = await response.json();
+            console.log(grade);
+
+            this.setState({ grade: grade, graded: true });
+            this.setState({ continue: grade['isCorrect'] });
+        } catch (e) {
+            console.log(e);
+        }
+    }
+
+    modalHandler = (event) => {
+        event.preventDefault();
+        this.setState({
+            showModal: !this.state.showModal,
+        });
```

Comment:
Check out the [React docs](https://reactjs.org/docs/react-component.html#setstate) on `setState` and [this article](https://codeburst.io/how-to-not-react-common-anti-patterns-and-gotchas-in-react-40141fe0dcd#828e) for why what's going on in line 129 here is an anti-pattern. In short: React's `setState` is asynchronous and optionally accepts a function (called an 'updater function' in the docs) that ensures that the state you want to update is what you expect to be at the moment of updating.

Reply:
I have updated how we setState to no longer depend on the current state based on the method shown in the article.